src/app_conn_pool.rs

```diff
@@ -0,0 +1,319 @@
+        &self,
+        installed_app_id: InstalledAppId,
+        admin_ws: AdminWebsocket,
+    ) -> HcHttpGatewayResult<AppWebsocket> {
+        {
+            let app_clients = self.app_clients.read().await;
+
+            if let Some(client) = app_clients.get(&installed_app_id) {
+                return Ok(client.app_ws.clone());
+            }
+        }
+
+        let mut app_client_lock = self.app_clients.write().await;
+
+        let app_ws = match app_client_lock.entry(installed_app_id.clone()) {
+            std::collections::hash_map::Entry::Occupied(client) => {
+                // Created by another thread while we were waiting for the lock
+                client.get().app_ws.clone()
+            }
+            std::collections::hash_map::Entry::Vacant(entry) => {
+                let app_ws = self
+                    .attempt_connect_app_ws(installed_app_id, admin_ws, 1)
+                    .await?;
+
+                entry.insert(AppWebsocketWithState {
+                    app_ws: app_ws.clone(),
+                    opened_at: Timestamp::now(),
+                });
+
+                app_ws
+            }
+        };
+
+        if app_client_lock.len() > self.configuration.max_app_connections as usize {
+            // Find and remove the oldest connection
+            let installed_app_id = app_client_lock
+                .iter()
+                .min_by_key(|(_, v)| v.opened_at)
+                .map(|(k, _)| k.clone())
+                .expect("Invalid lock");
+
+            tracing::warn!(
+                "Reached maximum app connections, removing connection for app: {}",
+                installed_app_id
+            );
+
+            app_client_lock.remove(&installed_app_id);
+        }
+
+        Ok(app_ws)
+    }
+
+    /// Remove an app client from the pool.
+    pub async fn remove_app_client(&self, installed_app_id: &InstalledAppId) {
+        self.app_clients.write().await.remove(installed_app_id);
+    }
+
+    async fn attempt_connect_app_ws(
+        &self,
+        installed_app_id: InstalledAppId,
+        admin_ws: AdminWebsocket,
+        retries: i8,
+    ) -> HcHttpGatewayResult<AppWebsocket> {
+        tracing::debug!(
+            "Attempting to connect to app client for {}",
+            installed_app_id
+        );
+
+        // Get the app port for a compatible app interface, which may be a cached value.
+        let app_port = self
+            .get_app_port(&installed_app_id, admin_ws.clone())
+            .await?;
+        tracing::debug!("Using app port {}", app_port);
+
+        // Issue an app authentication token to allow us to connect a new client.
+        let issued = admin_ws
+            .issue_app_auth_token(IssueAppAuthenticationTokenPayload::for_installed_app_id(
+                installed_app_id.clone(),
+            ))
+            .await?;
+
+        // Presence of host must have been checked to get an admin connection
+        let host = self
```

Comment:
This is thoroughly grim. This is why I was keen to store a `SocketAddr` on the configuration. I'd like to make that change as a follow up.

If you run this with `localhost` it will fail to make connections. It has to be an IP address at this point.